File: src/output/StopsWriter.py
from src.gtfs.models.StopData import StopData
from src.output.BinaryWriter import BinaryWriter


class StopsWriter(BinaryWriter):
    """Writer for stops.bin."""

    MAGIC = b"RST2"

    def write_header(self, schema_version: int, stop_count: int) -> None:
        """Write stops.bin header."""
        self.write_bytes(self.MAGIC)
        self.write_uint16(schema_version)
        self.write_uint32(stop_count)

    def write_stop(self, stop: StopData) -> int:
        """Write a single stop and return its offset."""
        stop_offset = self.offset

        self.write_uint32(stop.stop_id_internal)
        self.write_string(stop.name)
        self.write_float64(stop.lat)
        self.write_float64(stop.lon)

        # Write route references
        self.write_uint32(len(stop.route_ids))
        for route_id in stop.route_ids:
            self.write_uint32(route_id)

        # Write transfers
        self.write_uint32(len(stop.transfers))
        for target_stop, walk_time in stop.transfers:
            self.write_uint32(target_stop)
            self.write_int32(walk_time)

        return stop_offset

File: src/main/kotlin/com/raptor/output/StopsWriter.kt
package com.raptor.output

import com.raptor.gtfs.models.*
import java.io.OutputStream

class StopsWriter(stream: OutputStream) : BinaryWriter(stream) {
    private val magic = "RST2".toByteArray(Charsets.US_ASCII)

    fun writeHeader(schemaVersion: Int, stopCount: Int) {
        writeBytes(magic)
        writeUint16(schemaVersion)
        writeUint32(stopCount)
    }

    fun writeStop(stop: StopData): Long {
        val stopOffset = offset

        writeUint32(stop.stopIdInternal)
        writeString(stop.name)
        writeFloat64(stop.lat)
        writeFloat64(stop.lon)

        writeUint32(stop.route_ids.size)
        for (routeId in stop.route_ids) writeUint32(routeId)

        writeUint32(stop.transfers.size)
        for ((targetStop, walkTime) in stop.transfers) {
            writeUint32(targetStop)
            writeInt32(walkTime)
        }

        return stopOffset
    }
}
